GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -790,13 +798,33 @@ private void ClearProjectionCaches()
 
         private void RefreshSparseFolders()
         {
+            this.sparseFoldersNeedingRefresh = null;
             this.rootSparseFolder.Children.Clear();
             if (this.sparseCollection != null)
             {
+                this.sparseFoldersNeedingRefresh = new HashSet<string>(GVFSPlatform.Instance.Constants.PathComparer);
+
+                // Always refresh the root directory
+                this.sparseFoldersNeedingRefresh.Add(string.Empty);
+
                 Dictionary<string, SparseFolderData> parentFolder = this.rootSparseFolder.Children;
-                foreach (string directoryPath in this.sparseCollection.GetAll())
+
+                HashSet<string> oldSparsePaths = this.sparsePaths;
+                this.sparsePaths = this.sparseCollection.GetAll();
+                foreach (string directoryPath in this.sparsePaths)
                 {
                     string[] folders = directoryPath.Split(new[] { Path.DirectorySeparatorChar }, StringSplitOptions.RemoveEmptyEntries);
+                    if (oldSparsePaths != null && !oldSparsePaths.Remove(directoryPath))
+                    {
+                        // directoryPath is a new entry, add its parent to updatedFolders

```

Comment:
This might miss some required refresh scenarios.

*Example*

1. `gvfs sparse --set Scripts --prune`
2. `gvfs sparse --set Scripts;GVFS\GVFS --prune`

In this scenario we should refresh the root directory as well.
In this scenario